### PR TITLE
Load, parse and analyze dependencies in Analyzer, not entity finders.

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

File: src/analyzer.ts [hunk 1]
```diff
@@ -159,6 +171,7 @@ export class Analyzer {
     if (finders) {
       return findEntities(document, finders);
     }
+    return [];
```
Comment (rictic) [inline src/analyzer.ts]: We should take a look at turning on null tracking.

File: src/analyzer.ts [hunk 2]
```diff
@@ -109,12 +110,23 @@ export class Analyzer {
...
-    return new DocumentDescriptor(document, entities);
+    let dependencyDescriptors: Descriptor[] = entities
+        .filter((e) => e instanceof InlineDocumentDescriptor
+          || e instanceof ImportDescriptor);
+    let analyzeDependencies = dependencyDescriptors
+        .map((d) =>  {
+          if (d instanceof InlineDocumentDescriptor) {
+            return this.analyzeSource(d.type, d.contents, document.url);
+          } else if (d instanceof ImportDescriptor) {
+            return this.analyze(d.url);
```
Comment (rictic) [inline src/analyzer.ts]: Do we need to resolve this url relative to `d.url`?

File: src/ast/inline-document-descriptor.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+export class InlineDocumentDescriptor {
+  type: 'html' | 'javascript' | 'css' | string;
```
Comment (rictic) [inline src/ast/inline-document-descriptor.ts]: Minor improvement maybe:

```typescript
type: 'html' | 'javascript' | 'css' | /* etc... */string;
```

I'm big on selling typescript for our code as machine-checked documentation.

File: test/html/html-script-finder_test.js [hunk 4]
```diff
@@ -49,8 +49,7 @@ suite('HtmlScriptFinder', () => {
           assert.instanceOf(entities[0], ImportDescriptor);
           assert.equal(entities[0].type, 'html-script');
           assert.equal(entities[0].url, 'foo.js');
-          assert.instanceOf(entities[1], DocumentDescriptor);
-          assert.equal(entities[1].document.url, 'test.html');
+          assert.instanceOf(entities[1], InlineDocumentDescriptor);
```
Comment (rictic) [inline test/html/html-script-finder_test.js]:
```javascript
assert.equal(entities[1].type, 'js');
assert.equal(entities[1].contents, `console.log('hi')`);
```

File: test/analyzer_test.js [hunk 5]
```diff
@@ -116,8 +154,8 @@ suite('Analyzer', () => {
       })
       return analyzer.getEntities(document).then((entities) => {
         assert.equal(entities.length, 2);
-        assert.instanceOf(entities[0], DocumentDescriptor);
-        assert.instanceOf(entities[1], DocumentDescriptor);
```
Comment (rictic) [inline test/analyzer_test.js]: Consider using `chai-subset` to quickly assert a bunch of things in an easy to use way and future-unobtrusive way:

```javascript
expect(entities[0]).to.containSubset({type: 'js', contents: `console.log('hi')`});
expect(entities[1]).to.containSubset({type: 'css', contents: `body { color: red; }`});
```

**Comment by rictic** [comment]
Looks good save for the question about url resolution above, and a few recommendations on tests.

File: src/analyzer.ts (hunk 1 above)
Comment (justinfagnani) [inline src/analyzer.ts]: yes

File: src/analyzer.ts (hunk 2 above)
Comment (justinfagnani) [inline src/analyzer.ts]: The resolution is done in the entity finder... I suppose there's a question about where to do it, but I added to the test to make sure it's done.

File: src/ast/inline-document-descriptor.ts (hunk 3 above)
Comment (justinfagnani) [inline src/ast/inline-document-descriptor.ts]: done

File: test/html/html-script-finder_test.js (hunk 4 above)
Comment (justinfagnani) [inline test/html/html-script-finder_test.js]: done

File: test/analyzer_test.js (hunk 5 above)
Comment (justinfagnani) [inline test/analyzer_test.js]: considering! ;)

File: src/analyzer.ts (hunk 2 above)
Comment (rictic) [inline src/analyzer.ts]: +1

File: test/analyzer_test.js (hunk 5 above)
Comment (rictic) [inline test/analyzer_test.js]: 🤔💭

**Comment by rictic** [comment]
LGTM

**Pull request merged**